### PR TITLE
[Placement Group]Reschedule bundles when the node of bundles is dead

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
If a node is dead, the placement group scheduler should try to recover the group by rescheduling the bundles of the dead node. This should have higher priority than trying to place other placement groups.

## Related issue number
https://github.com/ray-project/ray/issues/9794

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.h:75]:
```suggestion
  /// Get the bundles of this placement_group (including unplaced).
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:180]: If I understand correctly, this is just the reverse index of node_to_leased_bundles_, is that right? Maybe this can be added as a comment.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:77]: Can you clarify in the comment that this is the location of existing bundles for this placement group? (I assume)

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:152]:
```suggestion
  /// Generate schedule context.
```

File: src/ray/protobuf/common.proto
Comment (ericl) [inline src/ray/protobuf/common.proto:147]: Do we really need this? It seems we can infer it based on whether the bundle exists on some node.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:153]: Can we avoid adding the `is_placed` boolean? It seems sufficient that a location is assigned to infer that a bundle is placed. In the GetUnplacedBundles() method we can check if the location is present for the bundle.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:45]: Since this is now handling partial scheduling, should we rename this function and update the comment? Perhaps it can be called `ScheduleUnplacedBundles()`.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:70]: IIUC `node_to_bundles` is never mutated during scheduling right? Can we make it a pointer to const to make this clear?

**Review by ericl — CHANGES_REQUESTED** [review]
Some comments. My main question is whether we can remove the `is_placed` flag, which seems brittle to me as it is duplicating state with the bundle locations tables.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:180]: Btw, I have almost the same map with a little different types. https://github.com/ray-project/ray/pull/10063 We should probably use the same type index for this.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:180]: Maybe we can merge this PR first and I can handle differences in my PR next week?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:72]: This method seems unnecessary.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:53]:
```suggestion
  auto &bundles = placement_group_table_data_.bundles();
```?

**Review by rkooo567 — COMMENTED** [review]
Can you also add Python tests?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ericl) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:180]: We could also pull this out into a separate data structure to avoid errors dealing with the multiple maps, but maybe that is better done in a separate PR.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:180]: @ericl Definitely +1. I was actually trying to do that in my PR, but the class was too shallow, so I reverted. Maybe after we merge those 2 PRs, it is a good time to do that. (We can make a class like `leasingContext`).

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:210]: (to be compatible to remove API)

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:128]: Good idea.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:150]: > When we reschedule, dont' we need to only rescheduled unplaced bundles?

Yeah, we only reschedule unplaced bundles, in `Schedule` function we get unplaced bundles to schedule.

**Comment by ffbin** [comment]
> Some comments. My main question is whether we can remove the `is_placed` flag, which seems brittle to me as it is duplicating state with the bundle locations tables.

Thanks, the `is_placed` flag is used to reschedule dead bundles after GCS restart. Currently, `Bundle` has no `address` field. if we add the address field, we can remove `is_placed` flag.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:128]: Actually, I have one concern about this rescheduling priority. If we have a placement group bundle that requires a unique resource (for example gpu resource when there’s only one gpu node), this can postpone creating all other placement group until a node with the resources is added. Is there any good solution to resolve this issue?

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:128]: Actually, I have one concern about this rescheduling priority. If we have a placement group bundle that requires a unique resource (for example gpu resource when there’s only one gpu node), this can postpone creating all other placement group until a node with the resources is added. Is there any good solution to resolve this issue?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:70]: fixed

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h]: fixed

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:215]: yeah, i have fix it, thanks.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:128]: `ResourceSet::IsSubset` increases the error code to return whether there is no such resource or the number of resources is insufficient. If all nodes do not have the required resources, the scheduling priority will be lowered(add a new PlacementGroupState `WAITING`), and the scheduling will be triggered when a new node is added.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc]: fixed

File: src/ray/protobuf/common.proto
Comment (ffbin) [reply src/ray/protobuf/common.proto]: I use `node_id` field instead of `is_placed` field. When GCS is restarted, we can use `node_id` to restore the bundle location.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:72]: We need to modify bundle by this function.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:72]: Ah, my bad. I thought it was the method for `gcsPlacementGroupManager`.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc]: This policy sounds good to me. 

cc @ericl

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Please wait for @ericl 's approval!

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:126]: Can you have a check here like? 
```
RAY_CHECK(placement_group->State() != CREATED | REMOVED)
```

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.h:159]: So, our priority will be 
RESCHEDULING
PENDING
WAITING

Is it correct?

**Review by ericl — APPROVED** [review]
Thanks, the code reads much more clear now.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.h:159]: yeah, it is correct.

**Comment by rkooo567** [comment]
Please merge the PR after this PR is merged https://github.com/ray-project/ray/pull/10063. I think both PR should be compatible, but if there's any issue, I can take a look at it in the near future (or if you can catch some of them here, it'll be great @ffbin!)

**Comment by ericl** [comment]
Java test failure looks unrelated

**Pull request merged**
